ext/ae_fast_decimal_formatter/ae_fast_decimal_formatter.c

```diff
@@ -0,0 +1,114 @@
+#include <ruby.h>
+#include <float.h>
+
+VALUE c_format_decimal(double num, int precision)
+{
+  char numstr[100];
+  char str[DBL_MAX_10_EXP + 2];
+  int numi, numlen, stri, strl;
+  int maxp = 5, i;
+  char formatstr[10];
+  int digits_to_comma;
+  double rounding_factor, rounded_num;
+
+  if (precision > maxp) { precision = maxp; }
+  if (precision < 0) { precision = 0; }
+
+  snprintf(formatstr, 10, "%%.%df", precision);
+  formatstr[9] = 0;
+
+  if (precision > 0) {
+    rounding_factor = pow(10.0, precision);
+    rounded_num = round(num * rounding_factor) / rounding_factor;
+  } else {
+    rounded_num = round(num);
+  }
+
+  if (precision > 0) { precision++; } // adjust for .
+
+  numlen = snprintf(numstr, 100, formatstr, rounded_num);
+  digits_to_comma = numlen - precision;
+  if (numstr[0] == '-') { digits_to_comma--; }
+
+  strl = numlen + (digits_to_comma / 3);
+  if ((digits_to_comma % 3) == 0) {
+    strl--;
+  }
+
+  if (strncmp(numstr, "-0.00000", strl) == 0) {
+    return rb_str_new("0.00000", 1 + precision);
+  }
+
+  stri = strl;
+  str[stri] = 0;
+  for (i = 1; i <= (precision + 1); i++) {
+    str[stri - i] = numstr[numlen - i];
+  }
+
+  stri -= (precision + 2);
+  numlen -= (precision + 1);
+  numi = 1;
+
+  while ((numlen - numi) >= 1) {
+    if ((numi % 3) == 0) {
+      str[stri] = ',';
+      stri--;
+    }
+    str[stri] = numstr[numlen - numi];
+    stri--;
+    numi++;
+  }
+
+  if (numstr[0] == '-') {
+    str[0] = '-';
+  } else {
+    if ((numi % 3) == 0) {
+      str[1] = ',';
+    }
+    str[0] = numstr[0];
+  }
+  return rb_str_new2(str);
+}
+
+struct ae_fast_decimal_formatter {
+  double num;
+  int precision;
+};
+
+static void ae_fast_decimal_formatter_free(void *p) {}
```

Comment:
Do we need to free the internal pointer?

Reply:
https://github.com/rubygems/guides/blob/173db855aeb25193616cc0f8eed2928f8bfa626d/ext/my_malloc/my_malloc.c#L22

Reply:
No need to free. We didn't malloc anything. This object in in stack, not in heap.

Reply:
Cool that makes sense.